text/0024-schema-syntax.md

```diff
@@ -0,0 +1,449 @@
+# Cedar Schema Syntax
+
+## Related issues and PRs
+
+- Reference Issues: 
+- Implementation PR(s): 
+
+## Timeline
+
+- Start Date: 2023-07-24
+- Date Entered FCP: 2023-08-22
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+This document proposes a custom syntax for Cedar schemas. The syntax was developed with the following goals in mind:
+
+* Use the same concepts in both the JSON-based and custom syntax, allowing them to be naturally inter-converted
+* Reuse Cedar policy syntax as much as possible, to help with intuition
+* When no Cedar policy syntax analogue exists, leverage ideas from other programming languages that favor readability
+* Use as few syntactic concepts as possible; similar ideas in the schema should look similar syntactically
+* Support converting back and forth between the JSON-based schema without losing information
+
+This is _not_ a proposal to _replace_ the JSON-based syntax; the aim is to provide an additional syntax that better satisfies the goals above. The JSON syntax is appropriate in many circumstances, such as when constructing schemas systematically. Adopting this proposal would put schemas on the same footing as Cedar policies, which have a custom and a JSON syntax.
+
+## Basic example
+
+Here is the proposed syntax for the [TinyTodo schema](https://github.com/cedar-policy/cedar-examples/blob/main/tinytodo/tinytodo.cedarschema.json), whose auto-formatted JSON version is 160 lines long.
+```
+entity Application;
+entity User in [Team,Application] { name: String };
```

Comment:
Suppose that we decide at some point to support named (transitively closed) relations, in addition to the (anonymous) hierarchy membership relation.  

What would be a nice way to extend the schema syntax to support these?

Reply:
shouldn't we first decide what the Cedar syntax would be, and then use that to inform the schema syntax?

Good chance it could be the same. for instance if (strawman) the syntax is `A in(mycustomrelationname) B`, then the schema could use the same

Reply:
Yes, I imagine we'd follow the Cedar syntax.   My question would be more precisely phrased as this:  do we see something about the current proposal that's a one-way door and would make such an addition particularly onerous?  (I don't but I haven't spent as much time thinking about this design as the other commenters :) )

Reply:
I don't see any issues, no. Basically: Whatever we can work into Cedar policy syntax _should_ work out Ok for schemas, since there is a lot of similarity between the two.